File: chp01_Singleton/Python/Testing_ConfigurationSettings_reset.py
import threading

class ConfigurationSettings:
    _instance = None
    _lock = threading.Lock()

    def __new__(cls):
        if cls._instance is None:
            with cls._lock:
                if cls._instance is None:
                    cls._instance = super(ConfigurationSettings, cls).__new__(cls)
        return cls._instance

    def get_config(self):
        return self.config

    def set_config(self, config):
        self.config = config

    @classmethod
    def reset_instance(cls):
        with cls._lock:
            cls._instance = None

# Test class using Dependency Injection
class ConfigurationSettingsTest:
    def __init__(self, configuration_settings):
        self.configuration_settings = configuration_settings

    def run_test(self):
        # Example test case: set and get configuration
        self.configuration_settings.set_config("Test Settings")
        config_value = self.configuration_settings.get_config()
        print(f"Configuration Value: {config_value}")

# Usage
config_settings1 = ConfigurationSettings()
config_settings2 = ConfigurationSettings()
config_settings1.set_config("Server Settings")
print(config_settings1 is config_settings2)  # Output: True 
print(config_settings1.get_config())  # Output: Server Settings
print(config_settings2.get_config())  # Output: Server Settings

# Dependency Injection Example
config_test = ConfigurationSettingsTest(config_settings1)
config_test.run_test()  # Output: Configuration Value: Test Settings

# Reset Singleton Instance for Testing Purposes
ConfigurationSettings.reset_instance()
config_settings3 = ConfigurationSettings()
print(config_settings1 is config_settings3)  # Output: False (since the singleton instance was reset)
config_settings3.set_config("New Server Settings")
print(config_settings3.get_config())  # Output: New Server Settings

File: chp01_Singleton/C++/Testing_ConfigurationSettings_reset.cpp
#include <iostream>
#include <string>
#include <mutex>

class ConfigurationSettings {
public:
    static ConfigurationSettings& getInstance() {
        // Use call_once to ensure the instance is created only once in a thread-safe manner
        std::call_once(initInstanceFlag, createInstance);
        return *instance;
    }

    static void resetInstance() {
        std::lock_guard<std::mutex> lock(instanceMutex);
        delete instance;
        instance = nullptr;
        std::cout << "ConfigurationSettings instance reset for testing purposes" << std::endl;
    }

    std::string getSetting(const std::string& key) {
        // Simulate getting a setting from a configuration file
        if (key == "godOfTheUnderworld") {
            return "Hades";
        }
        return "";
    }

private:
    ConfigurationSettings() {}
    ~ConfigurationSettings() {}

    // Declare but do not define copy constructor and assignment operator to prevent copying
    ConfigurationSettings(const ConfigurationSettings&);
    ConfigurationSettings& operator=(const ConfigurationSettings&);

    static void createInstance() {
        instance = new ConfigurationSettings();
    }

    static ConfigurationSettings* instance;
    static std::once_flag initInstanceFlag;
    static std::mutex instanceMutex;
};

ConfigurationSettings* ConfigurationSettings::instance = nullptr;
std::once_flag ConfigurationSettings::initInstanceFlag;
std::mutex ConfigurationSettings::instanceMutex;

// Test class using Dependency Injection
class ConfigurationSettingsTest {
public:
    ConfigurationSettingsTest(ConfigurationSettings& config) : configuration(config) {}

    void runTest() {
        std::string value = configuration.getSetting("godOfTheUnderworld");
        std::cout << "Test Configuration Value: " << value << std::endl;
    }

private:
    ConfigurationSettings& configuration;
};

int main() {
    ConfigurationSettings& settings = ConfigurationSettings::getInstance();
    std::cout << settings.getSetting("godOfTheUnderworld") << std::endl;

    // Dependency Injection Example
    ConfigurationSettingsTest configTest(settings);
    configTest.runTest();

    // Reset Singleton Instance for Testing Purposes
    ConfigurationSettings::resetInstance();
    ConfigurationSettings& settings2 = ConfigurationSettings::getInstance();
    std::cout << settings2.getSetting("godOfTheUnderworld") << std::endl;

    return 0;
}
